### PR TITLE
Setup associations

### PR DESCRIPTION
**Pull request by Tayyab-Kamboh**
Following tasks were completed:

In this PR following tasks are done.

-  Create a class Classroom
-  Create the has-many/belongs-to relationship between Classroom and Student
-  Create a class Book
-  Create a class Rental
-  Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental
-  Set up associations between classes and objects

### CODE REVIEW

File: classroom.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+  def initialize(label)
+    @label = label
+  end
+
+  def add_student(student)
+    @students << student
+    student.classroom = self
+  end
+end
```
Comment (anagudelogu) [inline classroom.rb]: - [x] You did a great job working on this milestone! 💯 However, in order to set up associations correctly, you need to implement the relationship in both ways. Here's a [refresher](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships) from a previous lesson. Currently, you can't assign a classroom to a student, you can only add students to the classroom. 🤓 🔧

File: rental.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  def initialize(date, book, person)
+    @date = date
+    @book = book
+    @person = person
+    person.rentals << self
+    book.rentals << self
+  end
+
+  attr_accessor :date, :book, :person
+end
```
Comment (anagudelogu) [inline rental.rb:11]: - [x] Here something similar happens, your `Person` class doesn't have an attribute called `rentals` so this code will fail.

```rb
# person.rb
require_relative 'nameable'
require_relative 'capitalize_decorator'
require_relative 'trimmer_decorator'


class Person < Nameable
  def initialize(age, name = 'Unknown', parent_permission: true)
    super()
    @id = Random.rand(1..1000)
    @name = name
    @age = age
    @parent_permission = parent_permission
  end


  def correct_name
    @name
  end


  attr_reader :id
  attr_accessor :name, :age, :parent_permission


  private


  def of_age?
    @age >= 18
  end


  public


  def can_use_services?
    of_age? || @parent_permission
  end
end
```

Please consider updating your person class in order to handle rentals. 🤓 🔧

**Review by anagudelogu — CHANGES_REQUESTED** [review]
# STATUS:  CHANGES REQUIRED  ♻️ 🚧 

Hi @Tayyab-Kamboh , 
I'm Andres.


While you have done a great job working on this milestone 💻 , there are some issues that you still need to work on to go to the next project. ♻️ 

You are almost there, keep up the good work! 🚀 



## Highlights
✅ Great use of Git flow. 💯 
✅ Your project is documented professionally. 📜 
✅ Descriptive PR title, summary and commit messages. 💬 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: student.rb [hunk 3]
```diff
@@ -2,8 +2,14 @@
...
-  def initialize(age, name = 'unknown', parent_permission: true, classroom: nil)
+  def initialize(name) # rubocop:todo Lint/MissingSuper
+    @name = name
+    @classroom = nil
+  end
+
+  def initialize(age, name = 'unknown', parent_permission: true, classroom: nil) # rubocop:todo Lint/DuplicateMethods
     super(name, age, parent_permission: parent_permission)
     @classroom = classroom
   end
```
Comment (anagudelogu) [inline student.rb]: You did a great job working on the changes requested 💯 , however, new bugs were introduced:

- [x] Please avoid disabling linter rules as they serve as a way of having consistent code formatting and ensuring best practices. 
- [x] Currently, you have your `:classroom` attribute repeated in your `attr_reader` and `attr_accessor` please avoid this repetition. 🤓 🔧 
- [x] You also have two initialize methods, please keep only one! 🤓 🔧 


- [x] Also, according to the previous review, the requirement:
> You did a great job working on this milestone! 100 However, in order to set up associations correctly, you need to implement the relationship in both ways. Here's a [refresher](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships) from a previous lesson. Currently, you can't assign a classroom to a student, you can only add students to the classroom. nerd_face wrench

Hasn't been implemented completely yet. Please make sure to fix this. 💡 Hint: create a custom `classroom=` setter method to achieve this, just like in the lesson provided. 🤓

File: student.rb [hunk 4]
```diff
@@ -12,3 +18,4 @@ def play_hooky
     '¯\(ツ)/¯'
   end
 end
+f
```
Comment (anagudelogu) [inline student.rb]: - [x] Please consider removing meaningless pieces of code, they make your codebase look ugly and hard to read! 🤓 🔧

**Review by anagudelogu — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️ 🚧 

Hi @Tayyab-Kamboh 👋🏼 
I'm Andres.

You've done a great job implementing some of the requested changes for this milestone 🚀, but there are still some which haven't been addressed yet. You're getting closer, keep up the good work! 🤖 🎆 



 ## Highlights
✅ Great use of Git flow. 💯 
✅ Your project is documented professionally. 📜 
✅ Descriptive PR title, summary and commit messages. 💬 


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by omar-muhamad — APPROVED** [review]
# Approved ✔
### Hi @Tayyab-Kamboh,

#### Great job, Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 

### Congratulations! 🎉 

giphy (1)

## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- N/A

## [Optional] suggestions:

- N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**